File: Dec2020/11Dec/HexadecimalNumbers.cpp
/*
Problem Name: Hexadecimal numbers
Problem Link: https://www.hackerearth.com/practice/algorithms/searching/linear-search/practice-problems/algorithm/yet-another-easy-problem-1f3273a0/
*/

#include<bits/stdc++.h>
using namespace std;

#define test(t)     int t; cin>>t; while(t--)

int gcd(int x, int fx){
    if (fx == 0)
        return x;
    return gcd(fx, x % fx); 
}

int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    test(t){
        int l, r, x=0;
        cin>>l>>r;
        
        for(int i=l; i<=r; ++i){
            int dec=i, fx=0;
            while(dec > 0){
                int r = dec % 16;
                fx += r;
                dec /= 16;
            }
            if(gcd(i, fx) > 1)
                x++;
        }
        cout<<x<<"\n";
    }
    return 0;
}

File: Dec2020/11Dec/Hexadecimalnumbers.java
/*
Problem Name: Hexadecimal numbers
Problem Link: https://www.hackerearth.com/practice/algorithms/searching/linear-search/practice-problems/algorithm/yet-another-easy-problem-1f3273a0/submissions/
*/
import java.util.*;
import java.io.*;
class TestClass {
    public static void main(String args[] ) throws Exception {
        Scanner in = new Scanner(System.in);
        PrintWriter out = new PrintWriter(System.out);
        int t = in.nextInt();
        for(;t>0;t--){
            int l = in.nextInt(), r = in.nextInt(), cnt = 0;
            for(;l<=r;l++){
                if(gcd(l,hex(l)) > 1)
                    cnt++;
            }
            out.println(cnt);
        }
        out.flush();
    }
    static int gcd(int a,int b){
        if(b == 0)
            return a;
        return gcd(b, a%b);
    }
    static int hex(int n){
        int s = 0;
        while(n > 0){
            s += n%16;
            n = n/16;
        }
        return s;
    }
}
